### PR TITLE
Add AppArmor validation logic

### PR DESCRIPTION
**Pull request by timstclair**
The validation checks the prerequisites described in the [AppArmor proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/apparmor.md#prerequisites).

In order to unblock the AppArmor implementation from waiting on the APIs to merge, this PR uses 2 helper stubs for handling the Pod API.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29812)

### CODE REVIEW

**Comment by timstclair** [comment]
Rebased and updated boilerplate (again!).

**Comment by timstclair** [comment]
Reassigning review to Dawn since I'm blocked on this.

File: pkg/security/apparmor/validate.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	if v.validateHostErr != nil {
+		return v.validateHostErr
+	}
+
+	loadedProfiles, err := v.getLoadedProfiles()
+	if err != nil {
+		return fmt.Errorf("could not read loaded profiles: %v", err)
+	}
+
+	for _, container := range pod.Spec.Containers {
```
Comment (dchen1107) [inline pkg/security/apparmor/validate.go]: If pod has initContainers, should we validate its profile too?

Comment (timstclair) [inline pkg/security/apparmor/validate.go]: Good catch! Done.

**Comment by timstclair** [comment]
@dchen1107 ping :)

File: pkg/security/apparmor/validate.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+	profileName := strings.TrimPrefix(profile, ProfileNamePrefix)
+	if !loadedProfiles[profileName] {
+		return fmt.Errorf("profile %q is not loaded", profileName)
+	}
+
+	return nil
+}
+
+func (v *validator) getLoadedProfiles() (map[string]bool, error) {
```
Comment (jfrazelle) [inline pkg/security/apparmor/validate.go]: I don't know if the format here is guaranteed

Comment (timstclair) [inline pkg/security/apparmor/validate.go]: I looked into it a bit, and couldn't find the profiles format documented anywhere (AppArmor documentation seems scarce), but I did find the [kernel implementation](https://github.com/torvalds/linux/blob/master/security/apparmor/apparmorfs.c#L748), and the [parsing](http://bazaar.launchpad.net/~apparmor-dev/apparmor/release-2.7/view/head:/utils/aa-status#L93) done by the `aa-status` tool. I updated the parsing to be a bit more precise.

**Comment by timstclair** [comment]
Thanks @jfrazelle

**Comment by jfrazelle** [comment]
LGTM

**Comment by dchen1107** [comment]
LGTM overall except one comment which should be addressed through doc, and outside this pr: 

Based on current proposed design, app profile is opaque to Kubernetes, including kubelet, which is good. But ubuntu apparmor has more features (more "hooks") than the upstream one which we are validating today through GCI image. But if the customers bring apparmor profiles from their ubuntu installations, in which case the profiles may not work correctly on GKE/GCI. This is a portability issue, and we should document it. 

Please squash the commits, we are ready to go.

**Comment by dchen1107** [comment]
@jfrazelle Thanks for reviewing this. I temporarily removed lgtm label so that @timstclair can have enough time to squash his commits. :-) Tim, please feel free to readd the label once you are done with squash.

**Comment by timstclair** [comment]
Squashed. Thanks all!

**Comment by jfrazelle** [comment]
oh so sorry!!

**Comment by timstclair** [comment]
Rebased.

**Pull request merged**